### ISSUE TITLE
[scheduler/controller-manager] Failed to update lock: net/http: request canceled

### ISSUE THREAD

**Issue by mac-chaffee** [kubernetes/kubernetes#107678]
### What happened?

If you shut down two out of three control plane nodes, there's a chance the remaining kube-scheduler and kube-controller-manager instances will log this line over and over:

```
E0120 20:42:53.703984       1 leaderelection.go:361] Failed to update lock: Put "https://sterling-cluster.k8s.renci.org:443/apis/coordination.k8s.io/v1/namespaces/kube-system/leases/kube-controller-manager?timeout=5s": net/http: request canceled (Client.Timeout exceeded while awaiting headers)
```

When the two control plane nodes are powered back on, those logs continue and neither kube-scheduler nor kube-controller-manager do any work during this time (such as start up critical daemonsets on the control plane nodes), causing requests to the kube-apiserver to error or timeout.

It looks like kube-scheduler and kube-controller-manager (which both use leaderelection from client-go) can get caught in an infinite loop where they keep trying to acquire the lock by using a dead client.

### What did you expect to happen?

Once etcd and kube-apiserver are fully operation after an outage, kube-scheduler and kube-controller-manager should automatically recover as well.

If leaderelection is failing for an extended period of time and no useful work is happening, I'd then expect kube-scheduler and kube-controller-manager to fail a health-check, which would cause the containers to restart. Instead, I had to restart them manually.

### How can we reproduce it (as minimally and precisely as possible)?

NOTE: This doesn't happen every time. I have yet to replicate the specific issue.

1. Run a 3-node cluster with kind and [calico as the CNI](https://alexbrand.dev/post/creating-a-kind-cluster-with-calico-networking/):
```
cat >> kind.yaml << EOF
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
  image: kindest/node:v1.21.2
- role: control-plane
  image: kindest/node:v1.21.2
- role: control-plane
  image: kindest/node:v1.21.2
networking:
  disableDefaultCNI: true
  podSubnet: 192.168.0.0/16
EOF

kind create cluster --config kind.yaml

kubectl create -f https://docs.projectcalico.org/manifests/tigera-operator.yaml
kubectl create -f https://docs.projectcalico.org/manifests/custom-resources.yaml
```
2. `docker pause` two nodes for a few minutes to simulate them being down
3. `docker resume` the two nodes. Check kube-scheduler/kube-controller-manager logs.

### Kubernetes version

<details>
```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.5", GitCommit:"aea7bbadd2fc0cd689de94a54e5b7b758869d691", GitTreeState:"clean", BuildDate:"2021-09-15T21:10:45Z", GoVersion:"go1.16.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.5", GitCommit:"aea7bbadd2fc0cd689de94a54e5b7b758869d691", GitTreeState:"clean", BuildDate:"2021-09-15T21:04:16Z", GoVersion:"go1.16.8", Compiler:"gc", Platform:"linux/amd64"}
```
</details>

### Cloud provider

<details>
Baremetal
</details>

### OS version

<details>

```console
$ cat /etc/os-release
NAME="Red Hat Enterprise Linux"
VERSION="8.4 (Ootpa)"
ID="rhel"
ID_LIKE="fedora"
VERSION_ID="8.4"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Red Hat Enterprise Linux 8.4 (Ootpa)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:8.4:GA"
HOME_URL="https://www.redhat.com/"
DOCUMENTATION_URL="https://access.redhat.com/documentation/red_hat_enterprise_linux/8/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 8"
REDHAT_BUGZILLA_PRODUCT_VERSION=8.4
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="8.4"

$ uname -a
Linux k8s-node16 4.18.0-305.25.1.el8_4.x86_64 #1 SMP Mon Oct 18 14:34:11 EDT 2021 x86_64 x86_64 x86_64 GNU/Linux
```
</details>

### Install tools

<details>
Kubespray 1.18
</details>

### Container runtime (CRI) and and version (if applicable)

<details>
containerd 1.4.9
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
Calico 3.19.2
</details>

**Comment by mac-chaffee** [comment]
/wg reliability

I think

**Comment by aojea** [comment]
are you able to reproduce it with kind?

if you do, please export the logs with `kind export logs` and attach them here in a tarball

**Comment by mac-chaffee** [comment]
I've still been unable to replicate the issue, but I found the root cause: If you have a validatingwebhook for the cluster such as [Gatekeeper](https://github.com/open-policy-agent/gatekeeper), then a partial cluster reboot can be unrecoverable without manual intervention. The cluster gets caught in a circular dependency like the following:

1. kube-controller-manager needs to boot up, so it tried to perform leader election with `PUT /leases` -->
2. kube-apiserver needs to invoke the validatingwebhook (gatekeeper) to verify the request -->
3. gatekeeper is unreachable (I think) because either kube-proxy wasn't running or Endpoints weren't being populated by controller-manager --> Return to step 1

Full details of the issue:
<details>
By looking at logs/metrics from around the time of the bug, kube-apiserver's 50th percentile response time for `PUT leases` (used by leaderelection) was 6-7 seconds, which is greater than the 5 second timeout period:
<img width="1375" alt="image" src="https://user-images.githubusercontent.com/7581860/150719774-724cb86d-d388-4aba-832c-486ddae9a614.png">

That was caused by connectivity issues to gatekeeper's validatingwebhook causing a pileup of slow requests across the cluster:

<img width="1725" alt="image" src="https://user-images.githubusercontent.com/7581860/150721316-ae293220-5b49-48e4-915c-908237f60bcb.png">

And that graph matches the number of running kube-proxy instances:

<img width="1725" alt="image" src="https://user-images.githubusercontent.com/7581860/150722245-9504a1b6-d5e7-43fd-ad4a-dc2ded973543.png">
</details>

So it seems kube-controller-manager and kube-scheduler were working as intended, but gatekeeper was the culprit. Maybe gatekeeper users could disable validation on `leases`, lower the timeout a lot more, or run more replicas of gatekeeper to ensure more instances are on healthy nodes.
